Sources/SwiftDocC/Utility/CollectionChanges.swift

```diff
@@ -0,0 +1,215 @@
+            // Removal within segment
+            let lowerSegmentCount  = removalIndex
+            let higherSegmentCount = segment.count - lowerSegmentCount - 1 // the 1 is for the removed element
+            
+            // Split the segment in two with a new removal segment in-between.
+            segments[0...0] = [
+                Segment(kind: .common, count: lowerSegmentCount),
+                Segment(kind: .remove, count: 1),
+                Segment(kind: .common, count: higherSegmentCount),
+            ]
+        }
+    }
+    
+    private func findSegment(toInsertAt index: Int) -> (segment: Segment, startOffset: Int, segmentIndex: Int)? {
+        // Insertions are applied in order. This means that we can start with the previous offset and index.
+        var offset = insertStartOffset
+        for segmentIndex in insertStartIndex ..< segments.count {
+            let segment = segments[segmentIndex]
+            if segment.kind == .remove {
+                continue
+            }
+            
+            if index <= offset + segment.count {
+                return (segment, offset, segmentIndex)
+            }
+            offset += segment.count
+        }
+        return nil
+    }
+    
+    mutating func insert(at insertIndex: Int) {
+        guard let (segment, startOffset, segmentIndex) = findSegment(toInsertAt: insertIndex) else {
+            assert(segments.count == 1 && segments[0].kind == .remove, """
+                The only case when a segment can't be found in the loop is if the only segment is a 'remove' segment.
+                This happens when all the 'common' elements are removed (meaning that the 'from' and 'to' values have nothing in common.
+                """)
+            
+            segments.append(Segment(kind: .insert, count: 1))
+            return
+        }
+        assert(segment.kind != .remove)
+        
+        insertStartOffset = startOffset
+        insertStartIndex  = segmentIndex
+        
+        guard segment.kind != .insert else {
+            segments[segmentIndex].count += 1
+            return
+        }
+        assert(segment.kind == .common)
+        
+        if insertIndex == startOffset {
+            // Insert at start of segment
+            segments.insert(Segment(kind: .insert, count: 1), at: segmentIndex)
+        } else if insertIndex == startOffset + segment.count {
+            // Insert at end of segment
+            let insertSegmentIndex = segmentIndex + 1
+            
+            // If this is the last segment, append a new 'insert' segment
+            guard insertSegmentIndex < segments.count else {
+                segments.append(Segment(kind: .insert, count: 1))
+                return
+            }
+            
+            switch segments[insertSegmentIndex].kind {
+            case .insert:
+                // If the next segment is an 'insert' segment, simply increment it
+                segments[insertSegmentIndex].count += 1
```

Comment:
Is this case possible? Since the documentation you linked above seems to say that insertions are then applied from lowest offset to highest?

Reply:
As long as this code is well behaved this shouldn't be possible. I'll turn it into an assertion.

Reply:
Updated in https://github.com/apple/swift-docc/pull/420/commits/77721d846f4644ad66b47d50f826000d1e1996e5